### PR TITLE
ACM HUB Deployment

### PR DESCRIPTION
**Pull request by prsurve**
Signed-off-by: prsurve <prsurve@redhat.com>

### CODE REVIEW

**Comment by vasukulkarni** [comment]
:+1:

**Review by vasukulkarni — DISMISSED** [review]

File: ocs_ci/deployment/deployment.py [hunk 1]
```diff
@@ -1068,6 +1075,72 @@ def patch_default_sc_to_non_default(self):
...
+
+        acm_hub_subscription_manifest = tempfile.NamedTemporaryFile(
+            mode="w+", prefix="acm_hub_subscription_manifest", delete=False
+        )
+        templating.dump_data_to_temp_yaml(
+            acm_hub_subscription_yaml_data, acm_hub_subscription_manifest.name
+        )
+        run_cmd(f"oc create -f {acm_hub_subscription_manifest.name}")
+        logger.info("Sleeping for 90 seconds after subscribing to ACM")
+        time.sleep(90)
```
Comment (shylesh) [inline ocs_ci/deployment/deployment.py:1122]: Instead of sleep can we do some polling kind of thing?

File: ocs_ci/deployment/deployment.py [hunk 2]
```diff
@@ -173,6 +173,13 @@ def deploy_cluster(self, log_cli_level="DEBUG"):
             setup_netsplit(
                 tmp_path, master_zones, worker_zones, x_addr_list, arbiter_zone
             )
+        if config.ENV_DATA["deploy_acm_hub_cluster"]:
```
Comment (shylesh) [inline ocs_ci/deployment/deployment.py:176]: I am bit unclear here, how do we know on which OCP cluster we want ACM to be deployed ?

Comment (Shrivaibavi) [reply ocs_ci/deployment/deployment.py:176]: +1

File: ocs_ci/deployment/deployment.py [hunk 3]
```diff
@@ -1068,6 +1075,72 @@ def patch_default_sc_to_non_default(self):
...
+            namespace=constants.ACM_HUB_NAMESPACE,
+        )
+        acm_mch.wait_for_resource(
+            condition="Running",
+            resource_name="multiclusterhub",
+            column="STATUS",
+            timeout=720,
+            sleep=5,
+        )
+        logger.info("MultiClusterHub Deployment Succeeded")
```
Comment (Shrivaibavi) [inline ocs_ci/deployment/deployment.py:1142]: do we also need an error/assert message if the ACM deployment failed in any previous steps like creation of subscription or package manifest or operatorgroup fails ?

File: conf/ocsci/acm_hub_deploy_2.4.yaml [hunk 4]
```diff
@@ -0,0 +1,5 @@
+---
+ENV_DATA:
+  deploy_acm_hub_cluster: true
+  acm_hub_channel: release-2.4
```
Comment (Shrivaibavi) [inline conf/ocsci/acm_hub_deploy_2.4.yaml:4]: is this going to be a constant one or do we need a placeholder for this too ?

Comment (prsurve) [reply conf/ocsci/acm_hub_deploy_2.4.yaml:4]: we will be creating yaml per acm release

File: ocs_ci/deployment/deployment.py (hunk 1 above)
Comment (prsurve) [reply ocs_ci/deployment/deployment.py:1122]: i guess we cannot as we dont have any status field in sub we are also doing same for ocs subscription install

File: ocs_ci/deployment/deployment.py (hunk 2 above)
Comment (prsurve) [reply ocs_ci/deployment/deployment.py:176]: i have added a condition where we deploy acm over 4.9 or later cluster

File: ocs_ci/deployment/deployment.py (hunk 3 above)
Comment (prsurve) [reply ocs_ci/deployment/deployment.py:1144]: wait for resource will raise the failure and we can get to know there was a failure

**Comment by petr-balogh** [comment]
Verification job which @jimdevops19 ran: https://ocs4-jenkins-csb-odf-qe.apps.ocp-c1.prod.psi.redhat.com/job/qe-deploy-multi-ocs-cluster-x3/205/console

File: ocs_ci/deployment/deployment.py [hunk 5]
```diff
@@ -173,6 +173,14 @@ def deploy_cluster(self, log_cli_level="DEBUG"):
...
                 tmp_path, master_zones, worker_zones, x_addr_list, arbiter_zone
             )
+        ocp_version = version.get_semantic_ocp_version_from_config()
+        if config.ENV_DATA["deploy_acm_hub_cluster"] and ocp_version >= version.VERSION_4_9:
+            try:
+                self.deploy_acm_hub()
+            except Exception as e:
+                logger.error(e)
+        else:
+            logger.warning("ACM HUB deployment is skipped")
```
Comment (petr-balogh) [inline ocs_ci/deployment/deployment.py:183]: I think that we don't need to inform about this skip for non ACM deployment.

File: ocs_ci/deployment/deployment.py [hunk 6]
```diff
@@ -1069,6 +1077,72 @@ def patch_default_sc_to_non_default(self):
...
+        run_cmd(
+            f"oc create -f {constants.ACM_HUB_MULTICLUSTERHUB_YAML} -n {constants.ACM_HUB_NAMESPACE}"
+        )
+        acm_mch = ocp.OCP(
+            kind=constants.ACM_MULTICLUSTER_HUB,
+            namespace=constants.ACM_HUB_NAMESPACE,
+        )
+        acm_mch.wait_for_resource(
+            condition="Running",
+            resource_name="multiclusterhub",
```
Comment (ebenahar) [inline ocs_ci/deployment/deployment.py:1139]: please save `multiclusterhub` in constants.py

File: ocs_ci/deployment/deployment.py [hunk 7]
```diff
@@ -1069,6 +1077,72 @@ def patch_default_sc_to_non_default(self):
...
+        logger.info("Creating MultiCluster Hub")
+        run_cmd(
+            f"oc create -f {constants.ACM_HUB_MULTICLUSTERHUB_YAML} -n {constants.ACM_HUB_NAMESPACE}"
+        )
+        acm_mch = ocp.OCP(
+            kind=constants.ACM_MULTICLUSTER_HUB,
+            namespace=constants.ACM_HUB_NAMESPACE,
+        )
+        acm_mch.wait_for_resource(
+            condition="Running",
```
Comment (ebenahar) [inline ocs_ci/deployment/deployment.py:1138]: please use ocs_ci/ocs/constants.py:85

Comment (prsurve) [reply ocs_ci/deployment/deployment.py]: Done

File: ocs_ci/deployment/deployment.py (hunk 6 above)
Comment (prsurve) [reply ocs_ci/deployment/deployment.py]: Done

File: ocs_ci/deployment/deployment.py (hunk 5 above)
Comment (prsurve) [reply ocs_ci/deployment/deployment.py]: Done

File: ocs_ci/deployment/deployment.py [hunk 8]
```diff
@@ -173,6 +173,14 @@ def deploy_cluster(self, log_cli_level="DEBUG"):
             setup_netsplit(
                 tmp_path, master_zones, worker_zones, x_addr_list, arbiter_zone
             )
+        ocp_version = version.get_semantic_ocp_version_from_config()
+        if config.ENV_DATA["deploy_acm_hub_cluster"] and ocp_version >= version.VERSION_4_9:
```
[the remark this answers is not in the document]
Comment (prsurve) [reply ocs_ci/deployment/deployment.py]: Done

**Review by petr-balogh — APPROVED** [review]

**Review by keesturam — APPROVED** [review]
Great work @prsurve 
LGTM!

File: ocs_ci/deployment/deployment.py [hunk 9]
```diff
@@ -173,6 +173,15 @@ def deploy_cluster(self, log_cli_level="DEBUG"):
             setup_netsplit(
                 tmp_path, master_zones, worker_zones, x_addr_list, arbiter_zone
             )
+        ocp_version = version.get_semantic_ocp_version_from_config()
+        if (
+            config.ENV_DATA.get("deploy_acm_hub_cluster")
+            and ocp_version >= version.VERSION_4_9
+        ):
+            try:
+                self.deploy_acm_hub()
```
Comment (shylesh) [inline ocs_ci/deployment/deployment.py:182]: @prsurve what I meant here earlier was, you are in a multicluter environment, so for ex: there are 3 cluster contexts are present, before invoking deploy_acm_hub, you need to switc the context.

Comment (prsurve) [reply ocs_ci/deployment/deployment.py:182]: yes you are right we need context switch here

Comment (prsurve) [reply ocs_ci/deployment/deployment.py:182]: i will add in next comment

Comment (prsurve) [reply ocs_ci/deployment/deployment.py:182]: i see ACM switch ctx is part of your pr hence we need your pr to be merged first

Comment (prsurve) [reply ocs_ci/deployment/deployment.py:208]: done

**Review by shylesh — APPROVED** [review]

**Comment by shylesh** [comment]
LGTM. This PR is a prerequisite for OCP deployment using ACM.

**Review by petr-balogh — APPROVED** [review]

**Comment by prsurve** [comment]
Verified Locally

**Pull request merged**